### DISCUSSION TITLE
Transform to MNI space - Visualization

### DISCUSSION THREAD

**Discussion by stavrizinonos** [TorchIO-project/torchio#586] [Q&A] [answered]
Hi,
I have segmented some brain MRI images, where i saved the initial MRI with the corresponding brain mask in a Subject object. Then, i tried to visualize them using the interactive plot volume, which seems to work fine. However, when i do transformations, it also works fine, until i transform it to MNI space. Then, i can no longer visualize the images as shown in the 4th screenshot. Why is that happening?  The same happens and when i use subject.plot() , i can see the visualization at first but after transformation no.  I am wondering how i can set starting points on the plots so i can view it when it is transformed to the new space.  Also, if i want to save the stripped image in my subject object as well, as a new parameter, shall i save it as a ScalarImage or LabelMap? I have tried both and i was also not able to visualize the stripped image not even before transformation, but when i try to plot it when it is not in the subject object it appears fine. Why is that happening again?
[Visualisation_brain.pdf](https://github.com/fepegar/torchio/files/6713231/Visualisation_brain.pdf)
Thank you in advance

**Comment by fepegar** [comment]
Hi, @stavrizinonos. I think your registration was not successful and the image is empty. Look at the `RuntimeWarning` you get. The code is [here](https://github.com/fepegar/miccai-educational-challenge-2019/blob/94067cdb99830ea21e51964ac9c20a34c28aaea4/visualization.py#L66-L75). Please take a look at those values you want to visualize.

**Reply by stavrizinonos** [reply]
Thanks for your reply. And i have tested that, I believe the image is not empty, because when i try different plotting i get something but of course not what i should get, but at least from this i can verify that the image is not empty. Also, the registration was applied to the whole dataset and i noticed that there is nothing wrong when the images come from the IXI dataset, and this problem only appears when the image comes from the FCP dataset. The only difference i saw and it might be related between those 2, is their initial orientation ,as in the IXI images initially had the PSR+ orientation, while the FCP ones come with LAS+ orientation. Is there a possibility this is related with the registration?
[Vis2.pdf](https://github.com/fepegar/torchio/files/6722667/Vis2.pdf)
This is an example of a non skull-stripped('anonymized') & a skull-stripped images from the FCP dataset:
[scan_mprage_skullstripped.nii.gz](https://github.com/fepegar/torchio/files/6722673/scan_mprage_skullstripped.nii.gz)
[scan_mprage_anonymized.nii.gz](https://github.com/fepegar/torchio/files/6722677/scan_mprage_anonymized.nii.gz)

**Reply by stavrizinonos** [reply]
@fepegar Hi, was there any progress regarding this issue? Sorry for messaging again, I am really struggling to understand what is happening and why. Is seems that orientation is not a problem as in when I try different LAS+ orientation images the registration works for some. However, for many images it works perfectly and for many not at all, as the pdf files i attached above show. Does this only happen to me? When trying any of the example images i attached the transformation to MNI work for others? 
Thank you!

**Reply by fepegar** [reply]
Hi. Sorry, I'm quite busy at the moment. Can you please share some code so I can reproduce the problem? Also, it could be great if you could post images and text in your comment, instead of attaching a PDF.

**Reply by stavrizinonos** [reply]
Yeah of course, sorry! 

The images that i used as inputs in this example are : 
[corr-Beijing_sub11072.nii.gz](https://github.com/fepegar/torchio/files/6766458/corr-Beijing_sub11072.nii.gz)
[corr-AnnArbor_sub00306.nii.gz](https://github.com/fepegar/torchio/files/6766459/corr-AnnArbor_sub00306.nii.gz)
[skull-stripped-IXI002-Guys-0828-T1.nii.gz](https://github.com/fepegar/torchio/files/6766460/skull-stripped-IXI002-Guys-0828-T1.nii.gz)

The code after the images are skull stripped and with corrected bias are then fed into images A and B and C and need to be transformed to MNI space as follows:

```python
#3 images
imageA = tio.ScalarImage('skull-stripped-IXI002-Guys-0828-T1.nii.gz')

imageB = tio.ScalarImage('corr-AnnArbor_sub00306.nii.gz')

imageC = tio.ScalarImage('corr-Beijing_sub11072.nii.gz')

#to_RAS function
to_ras = tio.ToCanonical()
imageA_ras = to_ras(imageA)
imageB_ras = to_ras(imageB)
imageC_ras = to_ras(imageC)
#to_mni function
mni = tio.datasets.Colin27()
to_mni = tio.Resample(mni.t1.path)
imageA_mni = to_mni(imageA_ras)
imageB_mni = to_mni(imageB_ras)
imageC_mni = to_mni(imageC_ras)
#plotting
imageA_mni.plot()
imageB_mni.plot()
imageC_mni.plot()
#before
print(imageA)
print(imageB)
print(imageC)
#after
print(imageA_mni)
print(imageB_mni)
print(imageC_mni)
```

Result is the following:

The 3 plots:
![to_mni_plot](https://user-images.githubusercontent.com/43760109/124523696-09215100-ddf0-11eb-8d7b-e6d2f979271e.png)

The details of the 3 images before and after transformation:
![to_mni_details](https://user-images.githubusercontent.com/43760109/124523711-19393080-ddf0-11eb-8582-a97b2190b206.png)

**Comment by romainVala** [comment, marked as answer]
Hello
I think that there is a confusion between coregistration to MNI and resampling to MNI.
The tio.Resample transform, is just a resampling ie a re-griding of the data to a given reference grid.

it is independant of a coregistration, which aim to realign 2 volume (event if they are differente grid (size, number of voxel). A affine coregistration can be achieve with traditional software, flirt from fsl, nifti_reg, ants, ect ..., Usually it compute a Affine.txt matrix  ( and if needed, a resampled version of the input volume moved to the reference volume (mni template in your case)

looking at your 3 volumes (in mrtrix viewer) one can see that corr-AnnArbor_sub00306.nii.gz is at a different spatial position (almost a FOV away) from the 2 others, I guess this is due to its affine, that is not well set.
So the solution here is to use a coregistration tool, that will compute the affine transformation to apply. If then you add this affine as argument of tio.resample (pre_affine_name) you should then reslice with the correct allignment (== coregistration)

**Comment by fepegar** [comment]
I agree with @romainVala. Please read carefully the docs for `Resample`, the [Fundamentals Concepts](https://simpleitk.readthedocs.io/en/master/fundamentalConcepts.html) docs from SimpleITK and [Coordinate Systems](https://nipy.org/nibabel/coordinate_systems.html) from NiBabel. There is a tool to register to MNI in [`resseg`](https://github.com/fepegar/resseg#episurg), you might be interested. We will probably include a transform to register to the MNI space soon (#428).

**Reply by romainVala** [reply]
> here is a tool to register to MNI in [`resseg`](https://github.com/fepegar/resseg#episurg), you might be interested.

I thing you put a wrong reference, ... ? this is not for registration

**Reply by fepegar** [reply]
Sure, that's a package for segmentation. But it includes a CLI tool for MNI registration using ANTS, it's the third line of the snippet in the link I shared.

For example I just tried this successfully:

```shell
conda create -n test python=3.8 ipython -y && conda activate test
pip install resseg antspyx
resseg-mni `python -c "import torchio as tio; print(tio.datasets.FPG().t1.path)"` -r fpg_mni.nrrd
tiohd -p fpg_mni.nrrd
```

![Figure_1](https://user-images.githubusercontent.com/12688084/124728224-50e6cc00-df07-11eb-8b96-336e463ebc57.png)

The registration took about 90 seconds.

**Reply by romainVala** [reply]
not even need to install ants locally !
so nice

**Reply by fepegar** [reply]
Well, it's installed in the second line: "`pip install resseg antspyx`".

**Reply by stavrizinonos** [reply]
Thanks both of you very much. Yeah now i get it , i thought 'Resample' will do the registration. I ended up using flirt instead.
Thanks for your help!

**Reply by lkshrsch** [reply]
how exactly does one use the resseg-mni function? I don't understand where the input data is specified in that line. Thanks!

**Reply by romainVala** [reply]
hi, 
I never used it, but it looks like the input data is the first argument of resseg-mni funtion
`python -c "import torchio as tio; print(tio.datasets.FPG().t1.path)"`  will indeed gives you the path of the image that will be coregister (and resample ) to mni

**Reply by fepegar** [reply]
@lkshrsch please take a look at https://linuxhint.com/mastering-backticks-linux-bash-scripts/ and https://github.com/fepegar/resseg/blob/ecf256c9ba8593c9fb2f33b184245125eb6cd91b/resseg/cli/resseg_mni.py#L8-L18
